### PR TITLE
Generate comparative gas repports on PR

### PR DESCRIPTION
**Pull request by Amxx**
Example visible in the #3531 test PR.

Capture d’écran du 2022-07-07 11-32-04

### CODE REVIEW

**Comment by Amxx** [comment]
FYI, it is normal that the test for this PR fails.

This is because the master branch doesn't yet produce any gas report, which this PR is trying to fetch as part of the gas comparaison.

File: scripts/checks/compareGasReports.js [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+const fs = require('fs');
+const chalk = require('chalk');
+const { argv } = require('yargs')
+  .env()
+  .options({
+    style: {
+      type: 'string',
+      choices: [ 'bash', 'markdown' ],
+      default: 'bash',
```
Comment (frangio) [inline scripts/checks/compareGasReports.js:11]: Please rename `bash` to `shell` or `terminal`.

File: scripts/checks/compareGasReports.js [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+    },
+  });
+
+// Deduce bas tx cost from the percentage denominator
```
Comment (frangio) [inline scripts/checks/compareGasReports.js:15]:
```suggestion
// Deduce base tx cost from the percentage denominator
```

File: scripts/checks/compareGasReports.js [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+      chalk[entry.max.delta > 0 ? 'red' : entry.max.delta < 0 ? 'green' : 'reset']((isNaN(entry.max.total) ? '-' : entry.max.total.toString()).padStart(8)),
+      chalk[entry.max.delta > 0 ? 'red' : entry.max.delta < 0 ? 'green' : 'reset']((isNaN(entry.max.delta) ? '-' : plusSign(entry.max.delta) + entry.max.delta.toString()).padStart(8)),
+      chalk[entry.max.delta > 0 ? 'red' : entry.max.delta < 0 ? 'green' : 'reset']((isNaN(entry.max.ratio) ? '-' : plusSign(entry.max.ratio) + entry.max.ratio.toFixed(2) + '%').padStart(8)),
+      chalk[entry.avg.delta > 0 ? 'red' : entry.avg.delta < 0 ? 'green' : 'reset']((isNaN(entry.avg.total) ? '-' : entry.avg.total.toString()).padStart(8)),
+      chalk[entry.avg.delta > 0 ? 'red' : entry.avg.delta < 0 ? 'green' : 'reset']((isNaN(entry.avg.delta) ? '-' : plusSign(entry.avg.delta) + entry.avg.delta.toString()).padStart(8)),
+      chalk[entry.avg.delta > 0 ? 'red' : entry.avg.delta < 0 ? 'green' : 'reset']((isNaN(entry.avg.ratio) ? '-' : plusSign(entry.avg.ratio) + entry.avg.ratio.toFixed(2) + '%').padStart(8)),
```
Comment (frangio) [inline scripts/checks/compareGasReports.js:138]: This is calling for a helper like `formatCells(entry.min)`. The same recommendation for the markdown formatter.

```js
const formatCells = (entry) => {
  const color = entry.delta > 0 'red' : entry.delta < 0 ? 'green' : 'reset';
  return [
    chalk[color](...),
    chalk[color](...),
    ...
}
```

```suggestion
      ...formatCells(entry.min),
      ...formatCells(entry.max),
      ...formatCells(entry.avg),
```

File: scripts/checks/compareGasReports.js [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        ))
+        .filter(([ current, previous ]) => current && previous)
+        .filter(([ current, previous ]) => !opts.hideEqual || JSON.stringify(current) !== JSON.stringify(previous))
+        .map(([ current, previous ]) => ({
+          contract,
+          method: current.method,
+          min: {
+            total: current.min,
+            delta: current.min - previous.min,
+            ratio: 100 * (current.min - previous.min) / (previous.min + BASE_TX_COST),
```
[the remark this answers is not in the document]
Comment (Amxx) [reply scripts/checks/compareGasReports.js]: yes

**Comment by Amxx** [comment]
Updated preview:

Capture d’écran du 2022-07-08 14-30-18

**Review by frangio — APPROVED** [review]

**Comment by htadashi** [comment]
@Amxx @frangio This is beautiful. Any chance to create an independent repository for this action? :)

**Comment by frangio** [comment]
That would be interesting but we will not be able to maintain an action for general use. Feel free to take this code and spin it off into an action.

**Pull request merged**